### PR TITLE
fix: Load Vaadin version from pom.properties instead of from vaadin-c…

### LINKED ISSUE #17017: Platform.getVaadinVersion() randomly returns Vaadin or Hilla version in hybrid apps
### Description of the bug

`Platform.getVaadinVersion()` reads the `vaadin-core-versions.json` file from classpath, parses out the `"platform"` string and returns it as the Vaadin version (e.g. 24.1.0). With Vaadin apps, the file is present in `vaadin-core.jar`.

The problem is that `hilla.jar` also contains the same `vaadin-core-versions.json` file, with `"platform"` string referring to Hilla version, e.g. 1.3.7 or 2.0.6. If both vaadin-core and hilla is present on the classpath in case of hybrid apps, this will lead to a problematic behavior.

Since `Platform.getVaadinVersion()` calls `Platform.class.getClassLoader().getResourceAsStream("vaadin-core-versions.json")`, it may parse either the file from `vaadin-core.jar` or from `hilla.jar`, whichever happens to be sooner on the classpath. That will lead to unpredictable results returned by the `Platform.getVaadinVersion()` function.

### Expected behavior

`Platform.getVaadinVersion()` should always return Vaadin version (e.g. 24.1.0), regardless of whether Hilla is or isn't on the classpath.

### Minimal reproducible example

Quite hard to reproduce unfortunately, requires a hybrid project with two different classpath orderings.

### Versions

- Vaadin / Flow version:
- Java version:
- OS version:
- Browser version (if applicable):
- Application Server (if applicable):
- IDE (if applicable):

### PR DESCRIPTION
**Pull request by mvysny**
…ore-versions.json

vaadin-core-versions.json file is present both in vaadin-core.jar and hilla.jar. For hybrid projects, the outcome of Platform.getVaadinVersion() is undeterministic. This causes the function to parse pom.properties file instead.

Fixes #17017

## Type of change

- [x] Bugfix
- [ ] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [ ] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

**Comment by caalador** [comment]
As a minimum it is missing tests for single vs multiple versions.
I'll leave the selection to marco as I made a competing fix.

**Comment by mvysny** [comment]
Probably a duplicate of https://github.com/vaadin/flow/pull/17034

**Comment by mcollovati** [comment]
@mvysny yours is also a good solution indeed, but I prefer the one by @caalador as it works on a file we own directly instead of the one generated by the build tool

**Comment by mvysny** [comment]
> @mvysny yours is also a good solution indeed, but I prefer the one by @caalador as it works on a file we own directly instead of the one generated by the build tool

@mcollovati yup that makes sense. The problem is that the `vaadin-core-versions.json` file may come from `vaadin-core.jar`, `vaadin-core-internal.jar` and/or `hilla.jar`, and it's quite hard to determine which one is the correct one. The problem is that in all three cases the file is located in the root directory, so you pretty much have to analyze the URL and perform some name matching of the parent jar file (or parent folder), to tell those apart. That of course only works if URLClassLoader is used, which may not be the case for some funky JavaEE servers. And of course the entire heuristics stops working for fatjars (if all jars are unpacked then packed into one fat jar).

The "load Maven pom.properties" solution works always since the pom.properties is stored under different folder within the jar file itself.

**Comment by mcollovati** [comment]
@mvysny good point!
It seems like that for Vaadin 24 the `vaadin-core-versions.json` is only in `vaadin-core-internal` and not in `vaadin-core`.
However, as mentioned also in the Mikael's PR, a better solution may be having a different file name for Hilla.

**Review by mcollovati — CHANGES_REQUESTED** [review]

File: flow-server/src/test/java/com/vaadin/flow/server/PlatformTest.java [hunk 1]
```diff
@@ -41,34 +42,64 @@ public class PlatformTest {
     public void rememberContextClassLoader() {
         oldContextClassLoader = Thread.currentThread().getContextClassLoader();
         Platform.hillaVersion = null;
+        Platform.vaadinVersion = null;
     }
 
     @After
     public void restoreContextClassLoader() {
         Thread.currentThread().setContextClassLoader(oldContextClassLoader);
```
[the remark this answers is not in the document]
Comment (mvysny) [reply flow-server/src/test/java/com/vaadin/flow/server/PlatformTest.java:48]: Done - I've created the `cleanMemoizedValues()` function which is called both before and after tests.

File: flow-server/src/main/java/com/vaadin/flow/server/Platform.java [hunk 2]
```diff
@@ -31,41 +30,49 @@
  * @since 23.0
  */
 public class Platform implements Serializable {
-
-    private static boolean versionErrorLogged = false;
-
+    private static final Logger log = LoggerFactory.getLogger(Platform.class);
```
[the remark this answers is not in the document]
Comment (mvysny) [reply flow-server/src/main/java/com/vaadin/flow/server/Platform.java]: Good catch, fixed

**Review by mcollovati — APPROVED** [review]

**Comment by mcollovati** [comment]
As discussed internally, we will proceed reading the version from `pom.properties`, but in the meanwhile we will work on renaming version JSON files and then revisit the `Platform` methods implementations.

**Pull request merged**